File: sketches/transfer-functions/tone.py
from scipy import signal
import numpy as np
import matplotlib.pyplot as plt
from enum import Enum

# Set the sample rate
sample_rate = 44100  # in Hz

# Change the tone value to see the difference in the frequency response
# Keep it between 0 and 1
tone = 0.75

def get_s_domain_coefficients(t):
    # The following transfer function was derived with QsapecNG:
    # (1-t)*R3 * C1 * C2 * R2 * R4 * V1 * s^2 + ( C1 * R4 * V1 * t*R3 + (1-t)*R3 * C1 * R4 * V1 + C1 * R2 * R4 * V1 + C1 * R1 * R4 * V1 + C1 * R1 * V1 * t*R3 ) * s + ( R4 * V1 + V1 * t*R3 )
    # ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
    # ( C1 * C2 * R2 * R4 * t*R3 + (1-t)*R3 * C1 * C2 * R2 * R4 + C1 * C2 * R1 * R2 * R4 + C1 * C2 * R1 * R2 * t*R3 + (1-t)*R3 * C1 * C2 * R1 * R2 ) * s^2 + ( C1 * R4 * t*R3 + (1-t)*R3 * C1 * R4 + C1 * R2 * R4 + C1 * R1 * R4 + C1 * R1 * t*R3 + (1-t)*R3 * C1 * R1 + C1 * R1 * R2 + C2 * R2 * R4 + C2 * R2 * t*R3 + (1-t)*R3 * C2 * R2 ) * s + ( R4 + t*R3 + (1-t)*R3 + R2 )

    # This function implements this transfer function, but with less repeated calculations.

    R1 = 2200
    R2 = 6800
    R3 = 20000
    R4 = 6800
    C1 = 2.2e-8
    C2 = 1e-7

    R3_a = t * R3
    R3_b = (1-t) * R3

    C1C2 = C1 * C2
    R2R4 = R2 * R4
    C1C2R2R4 = C1C2 * R2R4
    C1C2R1R2 = C1C2 * R1 * R2
    C1R4 = C1 * R4
    C1R1 = C1 * R1
    C2R2 = C2 * R2

    b0 = C1C2R2R4 * R3_a
    b1 = C1R4 * R3_b + C1R4 * R3_a + C1 * R2R4 + C1R1 * R4 + C1R1 * R3_b
    b2 = R4 + R3_b
    a0 = b0 + C1C2R2R4 * R3_b + C1C2R2R4 * R1 + C1C2R1R2 * R3_b + C1C2R1R2 * R3_a
    a1 = b1 + C1R1 * R3_a + C1R1 * R2 + C2R2 * R4 + C2R2 * R3_b + C2R2 * R3_a
    a2 = R4 + R3_b + R3_a + R2

    return ([b0, b1, b2], [a0, a1, a2])

# Get s-domain coefficients
num, den = get_s_domain_coefficients(tone)
print('s-domain coefficients', (num, den))

# Apply bilinear transform
b, a = signal.bilinear(num, den, fs=sample_rate)
print('z-domain coefficients', (list(b), list(a)))

# Get the frequency response
w,h = signal.freqz(b, a, 2**20)
w = w * sample_rate / (2 *np.pi)

# Plot the frequency response
fig1 = plt.figure(1)
plt.title('Digital filter frequency response')
plt.semilogx(w, 20 * np.log10(abs(h)), 'b')
plt.ylabel('magnitude [dB]')
plt.xlabel('frequency [Hz]')
plt.grid()
plt.axis('tight')
plt.xlim([10, 20000])
plt.ylim([-40, 0])
plt.show()

File: ds1/src/tone.rs
mod bilinear_transform;
mod biquad_filter;
use {bilinear_transform::BilinearTransform, biquad_filter::BiquadFilter};

const R1: f32 = 2200.;
const R2: f32 = 6800.;
const R3: f32 = 20000.;
const R4: f32 = 6800.;
const C1: f32 = 2.2e-8;
const C2: f32 = 1e-7;

pub struct Tone {
  bilinear_transform: BilinearTransform,
  filter: BiquadFilter,
}

impl Tone {
  const C1C2: f32 = C1 * C2;
  const R2R4: f32 = R2 * R4;
  const C1C2R2R4: f32 = Self::C1C2 * Self::R2R4;
  const C1C2R1R2: f32 = Self::C1C2 * R1 * R2;
  const C1R4: f32 = C1 * R4;
  const C1R1: f32 = C1 * R1;
  const C2R2: f32 = C2 * R2;

  pub fn new(sample_rate: f32) -> Self {
    Self {
      bilinear_transform: BilinearTransform::new(sample_rate),
      filter: BiquadFilter::new(),
    }
  }

  pub fn process(&mut self, input: f32, tone: f32) -> f32 {
    let s_domain_coefficients = Self::get_s_domain_coefficients(tone);
    let z_domain_coefficients = self.bilinear_transform.process(s_domain_coefficients);
    self.filter.process(input, z_domain_coefficients)
  }

  fn get_s_domain_coefficients(tone: f32) -> ([f32; 3], [f32; 3]) {
    let r3_a = tone * R3;
    let r3_b = (1. - tone) * R3;

    let b0 = Self::C1C2R2R4 * r3_a;
    let b1 =
      Self::C1R4 * r3_b + Self::C1R4 * r3_a + C1 * Self::R2R4 + Self::C1R1 * R4 + Self::C1R1 * r3_b;
    let b2 = R4 + r3_b;
    let a0 = b0
      + Self::C1C2R2R4 * r3_b
      + Self::C1C2R2R4 * R1
      + Self::C1C2R1R2 * r3_b
      + Self::C1C2R1R2 * r3_a;
    let a1 = b1
      + Self::C1R1 * r3_a
      + Self::C1R1 * R2
      + Self::C2R2 * R4
      + Self::C2R2 * r3_b
      + Self::C2R2 * r3_a;
    let a2 = R4 + r3_b + r3_a + R2;

    ([b0, b1, b2], [a0, a1, a2])
  }
}
